### PR TITLE
fix: transfer in `transferAsset`, `otcSend` and `otcClaim` to revert on silent failures

### PR DESCRIPTION
**Pull request by supercontracts**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved asset transfer handling to support tokens that return no data as well as tokens that explicitly return false, reducing failed transfers with non-standard tokens.
* **Tests**
  * Added test coverage and mock tokens for no-return-data success and explicit-false failure scenarios to validate transfer and OTC flows.

### CODE REVIEW

File: src/MainnetController.sol [hunk 1]
```diff
@@ -1168,6 +1150,35 @@ contract MainnetController is AccessControl {
...
+            asset,
+            abi.encodeCall(IERC20(asset).transferFrom, (source, destination, amount))
+        );
+
+        require(
+            returnData.length == 0 || abi.decode(returnData, (bool)),
+            "MainnetController/transfer-failed"
+        );
+    }
+
```
[the remark this answers is not in the document]
Comment (supercontracts) [reply src/MainnetController.sol:1181]: These are not needed in mintUSDS and burnUSDS because we know we are transferring USDS which returns true on a successful transfer

File: test/base-fork/TransferAsset.t.sol [hunk 2]
```diff
@@ -18,6 +18,117 @@ contract MockToken is ERC20 {
 
 }
 
+contract MockToken2 {
```
Comment (lucas-manuel) [inline test/base-fork/TransferAsset.t.sol]: Name this MocTokenkReturnNull, name the above one MockReturnFalse, and update both to be moved to a test/mocks/Mocks.sol file

File: test/mainnet-fork/OTCSwaps.t.sol [hunk 3]
```diff
@@ -27,6 +27,26 @@ contract ERC20 is ERC20Mock {
 
 }
 
+// Mock ERC20 with transfer returning false to simulate a failed transfer
+contract MockToken is ERC20 {
```
Comment (lucas-manuel) [inline test/mainnet-fork/OTCSwaps.t.sol]: Use Mocks.sol import here

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/mainnet-fork/OTCSwaps.t.sol (hunk 3 above)
Comment (supercontracts) [reply test/mainnet-fork/OTCSwaps.t.sol]: used

File: test/mocks/Mocks.sol [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    function _transfer(address owner_, address recipient_, uint256 amount_) internal {
+        balanceOf[owner_] -= amount_;
+
+        // Cannot overflow because minting prevents overflow of totalSupply,
+        // and sum of user balances == totalSupply.
+        unchecked { balanceOf[recipient_] += amount_; }
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (supercontracts) [reply test/mocks/Mocks.sol]: Not needed

File: test/base-fork/TransferAsset.t.sol (hunk 2 above)
Comment (supercontracts) [reply test/base-fork/TransferAsset.t.sol]: refactored

File: test/base-fork/TransferAsset.t.sol [hunk 5]
```diff
@@ -70,7 +59,7 @@ contract ForeignControllerTransferAssetFailureTests is TransferAssetBaseTest {
     }
 
     function test_transferAsset_transferFailed() external {
```
Comment (lucas-manuel) [inline test/base-fork/TransferAsset.t.sol]: `test_transferAsset_transferFailedOnReturnFalse`

File: test/base-fork/TransferAsset.t.sol [hunk 6]
```diff
@@ -1,22 +1,11 @@
...
-import { ERC20 } from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
-
 import { RateLimitHelpers } from "../../src/RateLimitHelpers.sol";
 
-import "./ForkTestBase.t.sol";
-
-contract MockToken is ERC20 {
-
-    constructor() ERC20("MockToken", "MockToken") {}
+import { MockTokenReturnFalse, MockTokenkReturnNull } from "../mocks/Mocks.sol";
```
Comment (lucas-manuel) [inline test/base-fork/TransferAsset.t.sol]: MockTokenReturnNull

File: test/mainnet-fork/OTCSwaps.t.sol [hunk 7]
```diff
@@ -429,17 +444,30 @@ contract MainnetControllerOTCClaimFailureTests is MainnetControllerOTCSwapBase {
...
+        vm.prank(relayer);
+        vm.expectRevert("MainnetController/transfer-failed");
+        mainnetController.otcClaim(exchange, address(token));
+    }
+
 }
 
 contract MainnetControllerOTCClaimSuccessTests is MainnetControllerOTCSwapBase {
 
     function test_otcClaim_usdt() external {
```
Comment (lucas-manuel) [inline test/mainnet-fork/OTCSwaps.t.sol:466]: Add note that this test covers the case of a ERC20 token returning null

Comment (lucas-manuel) [reply test/mainnet-fork/OTCSwaps.t.sol:466]: Also for octSend_usdt

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
